Add content cards documentation for Swift SDK

jerielng opened the pull request:
# Pull Request/Issue Resolution

#### Description of Change:
> Adding documentation for usage of Content Cards and Content Cards UI in the new Braze Swift SDK.

Closes #**SDK-3026**

#### Is this change associated with a Braze feature/product release?
- [X] Yes (**12/15/2022**)
- [ ] No

---

<details>
<summary>✔️ Pull Request Checklist</summary>
<br>

- [X] Check that you haven't removed any images (replacing an image with an updated one of the same name is fine), as this breaks the French site
- [ ] Check that all links work.
- [X] Ensure you have completed [our Contributors License Agreement](https://www.braze.com/docs/cla/).
- [ ] Tag @Timothy-Kim and @KellieHawks as a reviewer when your work is **done and ready to be reviewed for merge**. Are you an internal product manager? Reference the internal reviewing chart to tag the appropriate reviewer.
- [X] Tag others as reviewers as necessary.
- [ ] If you have modified any links, be sure to add redirects to `assets` > `js` > `broken_redirect_list.js`

</details>

<details>
<summary>⭐ Helpful Wiki Shortcuts</summary>
<br>

- [Writing Style Guide](https://docs.google.com/document/d/e/2PACX-1vTluyDFO3ZEV7V6VvhXE4As_hSFwmnFFdU9g6_TrAYTgH1QmbRoEDDdn5GzKAB9vdBbIdyiFdoaJcNk/pub)
- [Image Style Guide](https://docs.google.com/document/d/e/2PACX-1vRJSkwcjmjrTfLDagZccLpOMMyh5NN5SXRZSjz12cRAHbX4OrUmhvCmYpf_p5YB-9r4_jSOQLkicQIH/pub)
- [Styling Test Page](https://www.braze.com/docs/home/styling_test_page/)

</details>

<details>
<summary>❗ ATTN: For PR Reviewers</summary>
<br>

- [ ] Read our [Reviewing a PR page](https://github.com/Appboy/braze-docs/wiki/Reviewing-a-PR) for more on our reviewing suggestions.
- [ ] Read our [Previewing Documentation page](https://github.com/braze-inc/braze-docs/wiki/Previewing-and-Testing-Documentation) to see how to check the deployment.
  - [ ] Preview all changes in the linked Vercel environment by clicking the preview link in the vercel-bot comment in your PR.
</details>

<details>
<summary>❗ ATTN: Internal Reviewing Chart </summary>
<br>
<b>Work at Braze and not sure who to tag for review?</b> <br>Before tagging @timothy-kim or @KellieHawks for a general review, reference the following chart to see if a specific product vertical/reviewer applies to your pull request.
<br><br>
<table>
<tr>
    <td><b>Reviewer</b></td>
    <td><b>Product Vertical</b></td>
  </tr>
  <tr>
    <td>@josh-mccrowell-braze</td>
    <td>Monolith Deployments<br>Quality Infrastructure<br>Platform Infrastructure<br>Datalake<br>SDKs</td>
  </tr>
  <tr>
    <td>@kelliehawks</td>
    <td>Currents<br>Internal Tools<br>Product Partnerships<br>SMS<br>Customer Lifecycle, Identity and Permissions</td>
  </tr>
  <tr>
    <td>@bre-fitzgerald</td>
    <td>Reporting<br>Intelligence<br>User Targeting<br>IAM<br>Channels<br>FIX</td>
  </tr>
  <tr>
    <td>@lydia-xie</td>
    <td>Ingestion<br>Core Objects<br>Core Messaging<br>Messaging and Automation<br>Email (Composition and Infrastructure)</td>
  </tr>
</table>
</details>

jerielng commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/customization/customizing_feed.md:
```diff
@@ -0,0 +1,104 @@
+- Passing your `Attributes` struct to the `BrazeContentCardUI/ViewController` initializer.
+
+{% alert note %}
+Customization via `Attributes` is only available in Swift.
```
Double-checking if this is correct - I wasn't able to access the `Attributes` from Objective-C

hokstuff replied:
I vaguely remember ObjC customization was limited - sanity @lowip if that is true ^

lowip replied:
Yes, there's only Swift APIs available.

josh-mccrowell-braze replied:
This alert gave me pause, too. The only language samples in this article are in Swift. Is calling out that customization via `Attributes` needed if no other options are generally provided in this article?

jerielng commented:
@josh-mccrowell-braze The site preview seems to have failed during generation. I can't access the Vercel build to determine the root cause. Is this a known issue that can happen if the docs are written incorrectly?

josh-mccrowell-braze commented:
@jerielng Not super sure what went wrong, but it looks like it's working now. Please make sure I didn't introduce any errors into the work you were doing.

lowip commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/integration.md:
```diff
@@ -0,0 +1,152 @@
+
+## Card methods
+
+Each card can be initialized with a `Context` object, which contains various methods for managing your card's state. Call these methods when you want to modify the corresponding state property on a particular card object.
```
```suggestion
Each card is initialized with a `Context` object, which contains various methods for managing your card's state. Call these methods when you want to modify the corresponding state property on a particular card object.
```
Content cards created by the SDK always have a `context`.

lowip commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/integration.md:
```diff
@@ -0,0 +1,152 @@
+
+Each card can be initialized with a `Context` object, which contains various methods for managing your card's state. Call these methods when you want to modify the corresponding state property on a particular card object.
+
+|Method|Description|
+|---|---|
+| `logImpression` | Log the content card impression event. |
+| `logClick` | Log the content card click event. |
+| `processClickAction` | Process a given [`ClickAction`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/contentcard/clickaction) input. |
+| `logDismissed` | Log the content card dismissed event. |
+| `logError` | Log an error related to the content card. |
+| `loadImage` | Load a given content card image from a URL. This method can be nil when the content card does not have an image. |
+{: .reset-td-br-1 .reset-td-br-2}
```
Adds examples on how to call the methods.

```suggestion
| Method                               | Description                                                                                                                              |
|--------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
| `card.context?.logImpression()`      | Log the content card impression event.                                                                                                   |
| `card.context?.logClick()`           | Log the content card click event.                                                                                                        |
| `card.context?.processClickAction()` | Process a given [`ClickAction`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/contentcard/clickaction) input. |
| `card.context?.logDismissed()`       | Log the content card dismissed event.                                                                                                    |
| `card.context?.logError()`           | Log an error related to the content card.                                                                                                |
| `card.context?.loadImage()`          | Load a given content card image from a URL. This method can be nil when the content card does not have an image.                         |
{: .reset-td-br-1 .reset-td-br-2}
```

btw, I use the following extension in VSCode to prettify the tables:
```
Name: Markdown Table Prettifier
Id: darkriszty.markdown-table-prettify
Description: Transforms markdown tables to be more readable.
Version: 3.6.0
Publisher: Krisztian Daroczi
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=darkriszty.markdown-table-prettify
```

jerielng replied:
That's a good tip, thanks! It feels inefficient to keep opening a preview to double-check if everything was formatted correctly 😅

lowip commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/refreshing_the_feed.md:
```diff
@@ -0,0 +1,49 @@
+{% tabs %}
+{% tab Swift %}
+
+In Swift, Content Cards can be refreshed either with a completion handler or with an asynchronous return using the native Swift concurrency APIs.
```
```suggestion
In Swift, Content Cards can be refreshed either with an optional completion handler or with an asynchronous return using the native Swift concurrency APIs.
```

lowip reviewed (approved):
LGTM, see comments

hokstuff commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/customization/customizing_feed.md:
```diff
@@ -0,0 +1,104 @@
+
+Content Cards UI components are available in the `BrazeUI` library of the Swift SDK. You can customize your Content Cards UI elements and behavior by modifying the `Attributes` struct of `BrazeContentCardUI.ViewController`. Alternatively, the Content Card cells may also be subclassed and then used programmatically by overriding the `cells` property in the `Attributes` struct. Check out the [Examp ...
+
+It's also important to consider whether you should use a simple customization strategy with `Attributes` versus a completely custom view controller and [subscribe for data updates]({{site.baseurl}}/developer_guide/platform_integration_guides/swift/content_cards/integration/). For example, if you configure using `Attributes`, you can launch your Content Cards UI with minimal setup. However, if you ...
```
Optional: The last sentence is a bit long so maybe we can try to trim it?
```suggestion
It's also important to consider whether you should use a simple customization strategy with `Attributes` versus a completely custom view controller and [subscribe for data updates]({{site.baseurl}}/developer_guide/platform_integration_guides/swift/content_cards/integration/). For example, if you configure using `Attributes`, you can launch your Content Cards UI with minimal setup. However, if you use a complete view controller customization, you have more control over the card behavior, such as displaying in a carousel or adding interactive elements, but you would have to ensure impressions, dismissal events, and clicks are properly logged.
```

hokstuff commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/customization/customizing_feed.md:
```diff
@@ -0,0 +1,104 @@
+
+## Customizing UI
+
+The following code snippets show how to style and change out-of-the-Box Content Cards to fit your UI needs using methods provided by the SDK. These methods allow you to customize all aspects of the Content Card UI, including custom fonts, customized color components, customized text, and more. 
```
```suggestion
The following code snippets show how to style and change out-of-the-Box Content Cards to fit your UI using methods provided by the SDK. These methods allow you to customize all aspects of the Content Card UI, including custom fonts, customized color components, customized text, and more. 
```

hokstuff commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/integration.md:
````diff
@@ -0,0 +1,152 @@
+{% subtab CANCELLABLE %}
+```swift
+// This subscription is maintained through Braze cancellable, which will continue to observe for changes until the subscription is cancelled.
+let cancellable = AppDelegate.braze?.contentCards.subscribeToUpdates { [weak self] contentCards in
````
I think we can add wording [like in our public API](https://github.com/braze-inc/swift-sdk/blob/9a2acf4d02c81c0ebe38ef7ea9a536efbacc3a0d/Sources/BrazeKit/Api/ContentCards/ContentCardsApi.swift#L108) that you need a strong reference to this cancellable to keep it going

hokstuff reviewed (approved):
Small nits - LGTM!

josh-mccrowell-braze commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/customization/badges.md:
```diff
@@ -0,0 +1,58 @@
+## Requesting unread Content Card counts
+
+If you would like to display the number of unread Content Cards your user has, we suggest you request a card count and represent it with a Badge. Badges are a great way to call attention to new content awaiting your users in the Content Cards. If you'd like to add a badge to your Content Cards, you can filter the number of cards based on the `viewed` property.
+
```
I'm confused by these two sections. Is "unread Content Card counts" different from "unviewed Content Cards"? 

In the second section, I'm correct in understanding that these examples let you display a badge that displays the number of unread Content Cards to the end user, right?

lowip replied:
> Is "unread Content Card counts" different from "unviewed Content Cards"?

Nope, they're the same

> In the second section, I'm correct in understanding that these examples let you display a badge that displays the number of unread Content Cards to the end user, right?

Yes, like the notification badges on an app icon in the homescreen

josh-mccrowell-braze replied:
Gotcha, thanks! I'll edit for flow to make this a little clearer. 

Would it be possible to get a screenshot of a card with an unread number badge?

jerielng replied:
Aside from the unread indicator on a single content card, our UI doesn't display unread badges inside the app. This would typically be a badge on the app icon on the user's home screen of their phone, something like this: 
![image](https://user-images.githubusercontent.com/16450008/216636447-7fffa763-8bd2-4e0f-8623-25880ee27ed7.png)

I could generate one from one of our sample apps though - let me know what you think is best!

josh-mccrowell-braze replied:
Generating an image like that from our sample apps would be great! 

If you use a sample app, we might even provide specific line numbers for where to insert the sample code? Just a thought!

josh-mccrowell-braze commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/customization/customizing_feed.md:
```diff
@@ -0,0 +1,104 @@
+The following code snippets show how to style and change out-of-the-Box Content Cards to fit your UI using methods provided by the SDK. These methods allow you to customize all aspects of the Content Card UI, including custom fonts, customized color components, customized text, and more. 
+
+You can customize the default Braze Content Card UI by modifying the `Attributes` struct. BrazeUI offers two ways to configure this object:
+- Modifying the `Attributes.defaults` static variable.
```
Can we link to the Swift documentation for these methods here?

josh-mccrowell-braze commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/customization/handling_clicks_manually.md:
```diff
@@ -0,0 +1,46 @@
+
+# Handling clicks manually
+
+You can manually handle Content Card clicks by implementing the [`BrazeContentCardUIViewControllerDelegate`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazeui/brazecontentcarduiviewcontrollerdelegate) protocol and setting your delegate object as the `delegate` property of your `BrazeContentCardUI.ViewController`. Refer to the [Content Cards UI tutorial](https://braze-inc.github.io ...
```
What are some reasons a customer would want to handle content card clicks manually?

jerielng replied:
Typically, they would need to do this if they aren't leveraging the built-in Braze UI and are implementing their own UI logic. I'll go ahead and add this in!

josh-mccrowell-braze commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/customization/read_unread_indicators.md:
```diff
@@ -0,0 +1,62 @@
+
+## Customizing the unviewed indicator
+
+In general, the color of the unviewed indicator can be set alongside the other by assigning a color to the tint color of your `BrazeContentCardUI.ViewController` instance:
```
> "Can be set alongside the other..."

The other what? Can we clarify here?

jerielng replied:
I'll update this text to be more concise!

josh-mccrowell-braze replied:
Great rewrite!

josh-mccrowell-braze commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/customization/use_cases/carousel_view.md:
```diff
@@ -0,0 +1,61 @@
+---
```
If we only have the single use case, I'd recommend restructuring this section so it's not nested and is simply the last article under Customization. Nested sections are best for collecting similar articles together, and there's no need if there's only one.

If you agree, I can make this change.

josh-mccrowell-braze commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/customization/use_cases/carousel_view.md:
```diff
@@ -0,0 +1,61 @@
+- Building your own views
+- Logging Content Card analytics
+- Introducing additional client-side logic to dictate how many and which cards to show in the carousel
+
```
This is a great introduction, kudos.

josh-mccrowell-braze commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/refreshing_the_feed.md:
```diff
@@ -0,0 +1,49 @@
+
+---
+
+# Refreshing the feed
```
The content of this article is too small to stand on its own (typically, web content should be 500 words or more). What do you think about pulling this content into the integration article?

josh-mccrowell-braze commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/integration.md:
```diff
@@ -0,0 +1,154 @@
+
+Content Cards UI can be integrated from the `BrazeUI` library of the Swift SDK. This library provides two view controller contexts: navigation or modal. If you wish to intercept and react to the Content Card UI lifecycle, implement [`BrazeContentCardUIViewControllerDelegate`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazeui/brazecontentcarduiviewcontrollerdelegate) as the delegat ...
+
+### Navigation context
```
I'm not sure I understand the difference between a navigation or modal context. Unsure if this is something that is known by all of our intended users, but I think it would be worthwhile to provide more context about these contexts (haha, forgive me) to accommodate developer users of all experience levels.

hokstuff replied:
Maybe we can add a GIF to help highlight the differences, but it is an iOS concept

Modal comes from bottom with a separate element:
https://i.stack.imgur.com/2XEO4.gif

Navigation goes from side:
https://i.stack.imgur.com/1yNaF.gif

josh-mccrowell-braze replied:
Thanks for the clarification! The link helps a lot.

josh-mccrowell-braze commented on _docs/_developer_guide/platform_integration_guides/swift/content_cards/implementation_guide.md:
```diff
@@ -0,0 +1,94 @@
+
+## Sample use cases
+
+There are three sample customer use cases provided. Each use case offers a detailed explanation, relevant code snippets, and a look into how Content Card variables may look and be used in the Braze dashboard:
```
The intro says it offers relevant code snippets, but these are fairly light. Are there more example code that we can offer here? If not, I'd recommend cutting this intro down.

josh-mccrowell-braze commented:
Hi @jerielng ! I added some edits for style and flow. Would you please review these to make sure I did not change the meaning of the content and correctly interpreted the intentions of the original? I also added some one-off comments with some questions or suggestions.

jerielng commented:
> Hi @jerielng ! I added some edits for style and flow. Would you please review these to make sure I did not change the meaning of the content and correctly interpreted the intentions of the original? I also added some one-off comments with some questions or suggestions.

Hey @josh-mccrowell-braze - thanks for the feedback! I'll go through and update accordingly.

josh-mccrowell-braze reviewed (approved):
This looks fantastic. Thanks for everybody's hard work.
